### DISCUSSION TITLE
Karpor MCP Integration

### DISCUSSION THREAD

**Discussion by rajp152k** [KusionStack/karpor#675] [Ideas]
Discussions and Resources regarding the integration of Model Context Protocol into Karpor (towards v0.6) #658 

## Docs & Resources

 - https://spec.modelcontextprotocol.io/specification/
 - https://llmindset.co.uk/posts/2024/12/mcp-build-notes/

## Dev 

 - https://github.com/modelcontextprotocol/servers/tree/main/src/everything (sentinel debugging)
 - https://modelcontextprotocol.io/tutorials/building-mcp-with-llms

 ## Tooling 

 - https://github.com/mark3labs/mcphost
 - https://github.com/metoro-io/mcp-golang
 - https://github.com/mark3labs/mcp-go

## Server Index

 - https://github.com/punkpeye/awesome-mcp-servers?tab=readme-ov-file#databases

**Comment by rajp152k** [comment]
Hi @elliotxx @ffforest ,
given that MCP draws a whole bunch of parallels from LSP..

Karpor in this case will be analogous to the IDE (mcp-host (https://github.com/mark3labs/mcphost) can be used to build an adapter over karpor so that we can build in a decoupled manner..

The client-server interactions can be categorized into the given interface : 
https://spec.modelcontextprotocol.io/specification/#features

**Reply by rajp152k** [reply]
As for now, given the intentions of karpor to act as a consolidated APIServer and we're already pulling data into a karpor specific elasticdb and etcd (in the hub cluster): I think it's safe to say that MCP traffic will not cross the hub-work cluster boundary, but only the tool and function calls from the MCP server might do so in case in the future we decide to expand from just analytics to full blown cluster control..

please correct me if I am not thinking right..

**Reply by elliotxx** [reply]
@rajp152k I think there is no problem with your idea. We need to build Karpor-MCP-Server to implement MCP's interface. It can connect LLM and Karport-Storage (data source). Most requests will not directly access the k8s control plane to avoid putting pressure on it.

Once there is Karpor-MCP-Server, Karpor is like IDE. It can implement a Chat interface as the MCP-Client. Users can also use other clients (such as Claude-Desktop) to intelligently interact with data in Karpor.

**Reply by rajp152k** [reply]
built some context today (MCP + go)..  

was trying to setup a cli pipeline for quicker tests and iterations rather than relying on claude desktop..  

setup mcp-cli (https://github.com/chrishayuk/mcp-cli)  (allows for openai) and mcp-host (allows for claude) with ollama + mistral but function calling seems to be messed up  ( https://github.com/mark3labs/mcphost/issues/4 ) 

some more links and generic utils inserted for reference in the discussion header

getting closer to a satisfactory dev env upon which I can start iterating soon..  

the official inspector is something I wish to explore with depth tomorrow..

**Reply by elliotxx** [reply]
Really exciting execution 👍 Compared to Claude-Desktop, mcp-cli is obviously lighter and faster. And, I am a little confused about the role of mcphost. In the context of LSP, it is equivalent to [mason-lspconfig](https://github.com/williamboman/mason-lspconfig.nvim)?

**Reply by rajp152k** [reply]
mcp-cli would help with instantiating singular clients for interacting with a server;  
whereas mcp-host orchestrates multiple clients' interactions with multiple servers ..  

I think mason-lsp-config is doing more and across a larger span of time (from config to managing connections) compared to mcphost; I've used nvim and mason in the past but as black boxes and probably am missing out on the technicalities to answer accurately..  

I'm going ahead with mcp-cli for now though..

**Reply by rajp152k** [reply]
![image](https://github.com/user-attachments/assets/9f890fbf-8cc5-46a9-b985-3f501f43ff57)


from the docs: a host is like an ide that manages multiple clients

**Reply by elliotxx** [reply]
@rajp152k Yeah, It's much clearer now

**Reply by rajp152k** [reply]
using RAG over https://modelcontextprotocol.io/llms-full.txt with 4o-mini does somewhat good as an MCP Doc bot..  
see https://modelcontextprotocol.io/tutorials/building-mcp-with-llms  
will be adding some more golang context..

**Reply by elliotxx** [reply]
@rajp152k Amazing, this is truly a new era of development paradigm! Learned

**Reply by elliotxx** [reply]
@rajp152k Brother, have you encountered any problems recently

**Reply by rajp152k** [reply]
Hey, was busy for the weekend 
Am picking it up again now

**Reply by elliotxx** [reply]
@rajp152k Yeah, I'm looking forward to your thoughts and output! If there are any miscellaneous tasks, I am also willing to invest in the development of MCP

**Reply by rajp152k** [reply]
was able to have a cli-based interaction with an sqlite server with mcp-cli 

![image](https://github.com/user-attachments/assets/18df9a15-b649-492a-8e12-aff55c370032)

proceeding with a search for any existing servers written for elasticsearch and etcd now

**Comment by rajp152k** [comment]
hi @elliotxx  , I am now working on creating mcp-servers in golang for exposing karpor's etcd and elasticsearch and have a good idea of how to proceed..

I did tryout mcp-cli with mcp-go (I found mcp-golang to be a bit verbose so dropped that ), writing a basic calculator server https://github.com/mark3labs/mcp-go, building it, and using the build in the server-config for mcp-cli (https://github.com/chrishayuk/mcp-cli)

![image](https://github.com/user-attachments/assets/dfbc8b9c-3b2f-491c-a520-85a30217b4a4)

```bash
# in the mcp-cli directory
uv run mcp-cli --server calculator # calculator being the mcp-go server
# yields an interaction as one would expect
```

![image](https://github.com/user-attachments/assets/dd26742e-29f7-4600-9e32-9710c27d826e)


It works as expected and am progressing ahead..

rather than the usual npx/uvx we can just use go run if we wish to use a remote server down the line but use the build right away is sensible I guess

Some planning that I'll have to do is with regards to how I wish to categorize the interfaces with etcd and elasticsearch (resources, tools and prompts)..

am going to start with some basic CRUD and then iterate as karpor requires..

we can go for a unified server (for etcd and golang) rather than separate ones..

On the client side, @hai-tian , can you please see how the MCP SDK will fit into the webUI : I think this resource can be used for learning the generic behaviors of a client : https://modelcontextprotocol.io/quickstart/client

@elliotxx , I think I can start writing some code now; can you recommend how would you prefer to maintain the development process (branching strategy, maintaining sub tasks and issues?, recommendations other than the contributing docs (have read them), etc ? )

**Reply by rajp152k** [reply]
I'll also be starting with a deeper dive into how karpor manages its storage and see if I could reuse any of the existing development

**Reply by elliotxx** [reply]
Here are some documents that may be helpful to you 👇 If there is a missing content, please tell me at any time.
- https://www.kusionstack.io/karpor/developer-guide/contribution-guide/code-contribute
- https://www.kusionstack.io/karpor/developer-guide/conventions/release-process
- https://www.kusionstack.io/karpor/developer-guide/conventions/code-conventions
- https://www.kusionstack.io/karpor/developer-guide/conventions/commit-conventions

**Reply by elliotxx** [reply]
cc @hai-tian In the next version, we need to build a webui for MCP. You can read the following document in advance

**Reply by elliotxx** [reply]
@rajp152k Yo! Just shipped Karpor on PH - mind dropping by? 🙌 
https://www.producthunt.com/posts/karpor

**Reply by rajp152k** [reply]
upvoted, great stuff

**Comment by rajp152k** [comment]
hi @elliotxx, I wanted to better understand what the karpor api server is using etcd for

I have a grasp of how I can expose elasticsearch via an mcp-server (can map the actionables via `/home/rp152k/source/vcops/os-contribs/KusionStack/karpor/pkg/infra/search/storage/elasticsearch` 

but am not able to chart what exactly and how we can use etcd..
(am aware it uses it for the apiserver) but how should we proceed with exposing it (what endpoints,  etc..)

as for the mcp-client, I'll need help understanding how it fits in the UI and how should the server communicate with it..
as of now, I'm thinking of implementing the server as an stdio one rather than the SSE one..

Please let me know if any one else could pitch in on the design and how the components fit in

**Reply by elliotxx** [reply]
@rajp152k  Yeah, I was just about to contact you! and now I have entered the 0.7.0 version iteration, it is time to participate in the development of MCP with you.
- I have created an `mcp-support` branch that can serve as the development branch for this feature. Can you provide an initial PR? Let's start code review and development progress.
- Connecting etcd is a bit complicated (requires go-client for kubernetes). We can start with `elasticsearch` and step by step
- I understand that SSE is a relatively standard approach. We can discuss this point in depth :)
- Both @ruquanzhao and I can participate in discussions and development of MCP, while @hai-tian can complete the front-end work 🔥 You can assign tasks to us.

**Reply by rajp152k** [reply]
yes, I'll start working on that branch

**Reply by rajp152k** [reply]
@elliotxx , have opened a PR from mcp-support into main..

btw, once you close a PR/discussion/issue, it is only reactable/editable/commentable by collaborators and not contributors..  
so I'm not able to reply/react to you in the exact place when we conclude one of those..

**Reply by elliotxx** [reply]
@rajp152k Ops, I really didn't notice this. Maybe after the PR review is approved next time, you can click on merge. The same is true for issue, you can close it

**Comment by elliotxx** [comment]
@rajp152k  Hey, bro, I recently developed an mcp server for converting favicon (https://github.com/elliotxx/favicon-mcp-server), just for practice and to get familiar with the mcp development process. Is there any entry point for me to participate in the development?
